### DISCUSSION TITLE
How to work with a resource and a `httpLabel`

### DISCUSSION THREAD

**Discussion by zetashift** [disneystreaming/smithy4s#434] [Q&A] [answered]
Hiya,

I'm still new to Smithy and smithy4s. I'm trying to learn it by writing small services, one of these is a contacts list, just like a todo list...but with contacts heh.

My current model has a `resource : Contact` and it uses the id of a `Contact` in uri's, but this isn't possible because the resource identifier isn't allowed to have `@httpLabel`.

Am I modeling wrong, or is there a better/more standard way to do these type of things?

I have the following smithy file:

```kotlin
$version: "2"

namespace smithy4s.contacts

use smithy4s.api#simpleRestJson
use smithy4s.api#UUID

@simpleRestJson
service ContactsService {
  version: "0.0.1"
  operations: [AddContact, DeleteContact, ShowContact, ShowContacts]
}

@http(method: "POST", uri: "/contacts/{contactId}", code: 200)
operation AddContact {
  input: AddContactInput
  errors: [InvalidInput]
}

@idempotent
@http(method: "DELETE", uri: "/contacts/{contactId}", code: 200)
operation DeleteContact {
  errors: [InvalidInput]
}

@input
structure AddContactInput for Contact {
  @required
  $contactId
}

@readonly
@http(method: "GET", uri: "/contacts/{contactId}", code: 200)
operation ShowContact {
  output: ShowContactOutput
  errors: [InvalidInput]
}

@readonly
@http(method: "GET", uri: "/contacts", code: 200)
operation ShowContacts {
  output: ShowContactsOutput
}

@output
structure ShowContactsOutput {
  @required
  contacts: Contacts
}

@output
structure ShowContactOutput for Contact {
  @required
  $contactId
  
  name: String
}

resource Contact {
  identifiers: { contactId: ContactId }
}

list Contacts {
  member: ContactId
}

@error("server")
structure InvalidInput {
  @required
  message: String
}

@uuidFormat
string ContactId

```

**Comment by Baccata** [comment, marked as answer]
@zetashift, thanks for the interest :) 

For starters, let me say that in the current version, Smithy4s doesn't have any bespoke support for `Resource` whatsoever, so I obviously cannot give you a very informed piece of advice on resources. I've not ever used resources myself, and though we are planning on support it eventually, we haven't even started thinking about the "how" of a possible support. 

Now, that being said, let's try to minimise your example to make the problem clearer and identify a solution. This below seems to work. As you can see, I've added the `@httpLabel` annotation in the definition of AddContactInput

```
$version: "2"

namespace contacts

use smithy4s.api#simpleRestJson
use smithy4s.api#UUID
use smithy4s.api#uuidFormat


@simpleRestJson
service ContactsService {
  version: "0.0.1"
  operations: [AddContact]
}

@http(method: "POST", uri: "/contacts/{contactId}", code: 200)
operation AddContact {
  input: AddContactInput
}


@input
structure AddContactInput for Contact {
  @httpLabel
  @required
  $contactId
}


resource Contact {
  identifiers: { contactId: ContactId }
}


@uuidFormat
string ContactId
```

If you're worried about the repetition between various operations, you could use mixins to clear the boilerplate : 

```
@mixin
structure ContextInput for Contact {
  @httpLabel
  @required
  $contactId
}

structure AddContactInput with [ContextInput]{
}
```

**PS** please don't start your namespaces with `smithy4s`, we reserve the right to do weird things when namespaces start with `smithy4s. `

**Reply by zetashift** [reply]
It seems I had very wrong assumptions, sorry. I knew the docs mentioned that `Resource` of type shape are unsupported, but I interpreted like "oh so a certain type of resource is unsupported". 
AFAIK I don't have to use a resource, I used resources because all the smithy docs use it, and I'm trying to be more comfortable with the IDL. And the boilerplate was a concern, but praise mixin support in 2!

Also last wrong assumption was that the quickstart code uses the smithy4s namespace(`smithy4s.hello`) so I thought that namespace was necessary to active some codegen or something D:

One last small question, I think I can safely change to `Contact` to a `structure` right(using mixins to reduce the duplication)?

**Reply by Baccata** [reply]
> One last small question, I think I can safely change to Contact to a structure right(using mixins to reduce the duplication)?

I think so, for sure.

**Reply by zetashift** [reply]
Thank you very much! For the answer and the great library :D

**Comment by zetashift** [comment]
So I don't know if `structure` for `Contact` here is good/acceptable usage. Maybe `Contact` is better off being a resource? 
I don't have any specific requirements.


Currently I have this, after trying to clean up the code a bit(lil bit minified):

```kotlin
@simpleRestJson
service ContactsService {
  version: "0.0.1"
  operations: [AddContact, DeleteContact, ShowContact, ShowContacts]
}

@idempotent
@http(method: "DELETE", uri: "/contacts/{contactId}", code: 200)
operation DeleteContact {
  input: Contact,
  errors: [InvalidInput]
}

@http(method: "POST", uri: "/contacts/{contactId}", code: 200)
operation AddContact {
  input: AddContactInput
  errors: [InvalidInput]
}

@input
structure AddContactInput with [ContactContext] {
  @required
  $contactId
}

@mixin
structure ContactContext {
  @httpLabel
  @required
  contactId: ContactId
}

structure Contact {
  @httpLabel
  @required
  contactId: ContactId
}

@uuidFormat
string ContactId

```

Is this adequate? The `Delete` having a `Contact` as input, versus `AddContactInput` which uses a mixin, feels kinda funky.

**Reply by Baccata** [reply]
If you find it weird, or struggle to name things, you can use [inlined inputs](https://awslabs.github.io/smithy/2.0/guides/migrating-idl-1-to-2.html?highlight=inline#move-operation-inputs-and-outputs-inline) instead. Under the hood, it creates structures by taking the name of the operation and adding `Input` our `Output` suffixes. 

I do think in your case the name `Contact` fits the concept of resource indeed. But then again, I don't know yet what we'll do in Smithy4s with resources so 🤷

**Reply by zetashift** [reply]
Thank you, it's not really that I find it weird, I just don't know what's "normal", since smithy4s doesn't support `resources` which is fine, but what is the idiomatic way to do this CRUD type of things. It gets a bit harder to discover as I am experimenting and learning at the same time :P.
However the inlined inputs are very nice, I applied them to my code, and removed the `list` thing because that might be too hard for me now, and pagination I'll work out once I get past this phase.

```kotlin
$version: "2"

namespace zetashift.contacts

use smithy4s.api#simpleRestJson
use smithy4s.api#uuidFormat

@simpleRestJson
service ContactsService {
  version: "0.0.1"
  operations: [AddContact, DeleteContact, GetContact]
}

@http(method: "POST", uri: "/contacts", code: 200)
operation AddContact {
  input := {
    name: String
    description: String
  },
  output := {
    contactId: ContactId
  },
  errors: [InvalidInput]
}

@idempotent
@http(method: "DELETE", uri: "/contacts/{contactId}", code: 200)
operation DeleteContact {
  input := {
    contact: Contact
  },

  output := {},

  errors: [InvalidInput]
}

@readonly
@http(method: "GET", uri: "/contacts/{contactId}", code: 200)
operation GetContact {
  input := {
   contact: Contact
  },
  output := {
    contactId: ContactId,
    name: String,
    description: String
  },
  errors: [InvalidInput]
}

structure Contact {
  @required
  @httpLabel
  contactId: ContactId

  name: String
  description: String
}

@error("server")
structure InvalidInput {
  @required
  message: String
}

@uuidFormat
string ContactId

```
Is my current code, but I'm getting the following error:
```sh
[error] (core / Compile / smithy4sCodegen) software.amazon.smithy.model.validation.ValidatedResultException: Result contained ERROR severity validation events:
[error] [DANGER] zetashift.contacts#DeleteContact: This operation uses the `DELETE` method in the `http` trait, but the following members are sent as part of the payload of the request: `contact`. These members are sent as part of the payload because they are not explicitly configured to be sent in headers, in the query string, or in a URI segment. Many HTTP clients do not support payloads with DELETE requests. Consider binding these members to other parts of the HTTP request such as query string parameters using the `httpQuery` trait, headers using the `httpHeader` trait, or URI segments using the `httpLabel` trait. | HttpMethodSemantics /home/rishi/dev/scala/friendslist/modules/core/src/main/smithy/FriendsService.smithy:27:1
[error] [DANGER] zetashift.contacts#GetContact: This operation uses the `GET` method in the `http` trait, but the following members are sent as part of the payload of the request: `contact`. These members are sent as part of the payload because they are not explicitly configured to be sent in headers, in the query string, or in a URI segment. Many HTTP clients do not support payloads with GET requests. Consider binding these members to other parts of the HTTP request such as query string parameters using the `httpQuery` trait, headers using the `httpHeader` trait, or URI segments using the `httpLabel` trait. | HttpMethodSemantics /home/rishi/dev/scala/friendslist/modules/core/src/main/smithy/FriendsService.smithy:39:1
[error] [ERROR] zetashift.contacts#DeleteContact: This operation uses `zetashift.contacts#DeleteContactInput` as input, but the following URI labels found in the operation's `http` trait do not have a corresponding member marked with the `httpLabel` trait: `contactId` | HttpLabelTrait /home/rishi/dev/scala/friendslist/modules/core/src/main/smithy/FriendsService.smithy:28:1
[error] [ERROR] zetashift.contacts#GetContact: This operation uses `zetashift.contacts#GetContactInput` as input, but the following URI labels found in the operation's `http` trait do not have a corresponding member marked with the `httpLabel` trait: `contactId` | HttpLabelTrait /home/rishi/dev/scala/friendslist/modules/core/src/main/smithy/FriendsService.smithy:40:1
[error] [SUPPRESSED] smithy4s.api#UUID: The string shape is not connected to from any service shape. (This is a library namespace.) | UnreferencedShape jar:file:/home/rishi/.cache/coursier/v1/https/repo1.maven.org/maven2/com/disneystreaming/smithy4s/smithy4s-protocol/0.15.2/smithy4s-protocol-0.15.2.jar!/META-INF/smithy/smithy4s.smithy:15:1
[error] Total time: 0 s, completed Sep 4, 2022, 12:51:34 AM
```

I feel like I'm struggling with `httpLabel` the most, it can only be put on structure members, so I can't attach it to inlined definitions in `input`. But here I have it in a structure and it still fails

**Reply by Baccata** [reply]
Here's a corrected spec : 

``` 
$version: "2"

namespace zetashift.contacts


service ContactsService {
  version: "0.0.1"
  operations: [AddContact, DeleteContact, GetContact],
  errors: [InvalidInput]
}

@http(method: "POST", uri: "/contacts", code: 200)
operation AddContact {
  input := {
    name: String
    description: String
  },
  output := {
    contactId: ContactId
  }
}

@idempotent
@http(method: "DELETE", uri: "/contacts/{contactId}", code: 200)
operation DeleteContact {
  input := with[ContactInput] {},
}

@readonly
@http(method: "GET", uri: "/contacts/{contactId}", code: 200)
operation GetContact {
  input := with[ContactInput]{},
  output: Contact
}

@mixin
structure ContactInput {
  @required
  @httpLabel
  contactId: ContactId
}

string ContactId
structure Contact {
  name: String
  description: String
}

@error("client")
structure InvalidInput {
  @required
  message: String
}

```

> But here I have it in a structure and it still fails 

So, the input and output of an operation MUST be structures, and the httpLabel will only be mapped to path parameters when it is applied on a top-level member in that structure. 

When you do the following : 

``` 
@readonly
@http(method: "GET", uri: "/contacts/{contactId}", code: 200)
operation GetContact {
  input := {
   contact: Contact
  }
  ... 

```

.... what you're modelling is essentially 

```
def getContact(input: GetContactInput) : ...
case class `GetContactInput(contact: Contact)` 
case class `Contact(@httpLabel contactId: ContactId, ...)`
```

Because the `@httpLabel` annotation is applied on a nested member, smithy complains. It is simply not reconciling http-bindings in nested data structures with the requirements. So basically what I've done is make `ContactInput` a mixin (which is basically akin to a Scala trait), and apply that mixin on the inlined input. 

I've also made the `InvalidInput` a client error (ie the kind of error that fall into the 4xx range of http status codes), and moved it to the `service` definition to reduce the repetition.

**Reply by Baccata** [reply]
PS : I **highly** recommend you to install vscode and one of the Smithy extensions for it, either the one from [AWS](https://marketplace.visualstudio.com/items?itemName=Smithy.smithy-vscode-extension) or [our own](https://marketplace.visualstudio.com/items?itemName=disneystreaming.smithy). The two use the same LSP implementation (we implemented the first few iterations and have donated it to AWS, and we might deprecate our extension in the future as a result). 

AWS is developing an Intellij plugin, but until it's there, vscode can give you a very nice editor experience for editing smithy files.

**Reply by zetashift** [reply]
Ah thank you so much!! I didn't know you could put mixins in inline input definition.
I hope I get better at mastering this Smithy IDL x).

So when you have something like:
```kotlin
@http(method: "POST", uri: "/{name}", code: 200)
operation Hello {
  input: Person
  output: Greeting
}

structure Person {
  @httpLabel
  @required
  name: String
```
`Person` is seen as a non nested member?

**Reply by zetashift** [reply]
Ah yea I already use the Smithy extension, the extension + Smithy errors(very nice errors !), helped me avoid having a huge amount of "
how do I do this" moments, and ask you more dumb questions :P.

But since there are so many things you can do with Smithy, and with the new-ness of this library, I am mostly unsure of the "idiomatic" ways to do things, also because I only know a little about everything

**Reply by Baccata** [reply]
> So when you have something like:
> Person is seen as a non nested member?

Yup.
